runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -262,41 +272,190 @@ public long getConflictSetAndCommit(ConflictSetInfo conflictSet) {
+                    .append(
+                            affectedStreams,
+                            collectWriteSetEntries(),
+                            new TxResolutionInfo(getTransactionID(),
+                                    getSnapshotTimestamp(),
+                                    hashedConflictSet,
+                                    getWriteSetInfo().getHashedConflictSet())
+                    );
+        } catch (TransactionAbortedException tae) {
+            // If precise conflicts aren't required, re-throw the transaction.
+            if (!builder.isPreciseConflicts()) {
+                throw tae;
+            }
 
-                        // TxResolution info:
-                        // 1. snapshot timestamp
-                        // 2. a map of conflict params, arranged by streamID's
-                        // 3. a map of write conflict-params, arranged by
-                        // streamID's
-                        new TxResolutionInfo(getTransactionID(),
-                                getSnapshotTimestamp(),
-                                conflictSet.getHashedConflictSet(),
-                                getWriteSetInfo().getHashedConflictSet())
-                );
+            // Otherwise, do a precise conflict check.
+            address = doPreciseCommit(tae, conflictSet.getConflicts()
+                    , hashedConflictSet, affectedStreams);
+        }
 
-        log.trace("Commit[{}] Acquire address {}", this, address);
+        log.trace("doCommit[{}] Acquire address {}", this, address);
 
         super.commitTransaction();
         commitAddress = address;
 
         tryCommitAllProxies();
-        log.trace("Commit[{}] Written to {}", this, address);
+        log.trace("doCommit[{}] Written to {}", this, address);
         return address;
     }
 
+    /** Do a precise commit, which is guaranteed not to produce an abort
+     * due to a false conflict. This method achieves this guarantee by scanning
+     * scanning the log when the sequencer detects a conflict, and manually
+     * inspecting each entry in the conflict window. If there are no true conflicts,
+     * the transaction is retried with the sequencer, otherwise, if a true conflict
+     * is detected, the transaction aborts, with a flag indicating the the abort
+     * was precise (guaranteed to not be false).
+     *
+     * @param originalException     The original exception when we first tried to commit.
+     * @param conflictSet           The set of objects this transaction conflicts with.
+     * @param hashedConflictSet     The hashed version of the conflict set.
+     * @param affectedStreams       The set of streams affected by this transaction.
+     * @return                      The address the transaction was committed to.
+     * @throws TransactionAbortedException  If the transaction must be aborted.
+     */
+    protected long doPreciseCommit(@Nonnull final TransactionAbortedException originalException,
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 23 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)